File: src/main/java/rest/InvitacionResource.java
package rest;

import dao.DaoInvitacion;
import entity.Invitacion;
import org.hibernate.validator.constraints.NotEmpty;
import usecase.InvitacionUseCase;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/invitaciones")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class InvitacionResource {

    private final DaoInvitacion daoInvitacion;
    private final InvitacionUseCase invitacionUseCase;

    public InvitacionResource(DaoInvitacion daoInvitacion, InvitacionUseCase invitacionUseCase){
        this.daoInvitacion = daoInvitacion;
        this.invitacionUseCase = invitacionUseCase;
    }

    @POST
    public Response hacerInvitacion(@Valid @NotNull Invitacion invitacion){
        Object nuevaInvitacion = invitacionUseCase.hacerInvitacion(invitacion);
        return nuevaInvitacion != null ?
                Response.ok(nuevaInvitacion).build() :
                Response.status(Response.Status.BAD_REQUEST).build();
    }

    @PUT
    @Path("/{idInvitacion}")
    public Response responderInvitacion(@PathParam("idInvitacion") String idInvitacion,
                                        @Valid
                                        @NotNull Invitacion invitacion,
                                        @QueryParam("aceptar")
                                        @NotEmpty
                                        @NotNull Boolean acepto){
        boolean operacionExitosa;
        if(acepto){
            operacionExitosa = daoInvitacion.aceptarInvitacion(invitacion, idInvitacion);
        }else{
            operacionExitosa = daoInvitacion.rechazarInvitacion(invitacion, idInvitacion);
        }
        return operacionExitosa ?
                Response.ok().build() :
                Response.status(Response.Status.BAD_REQUEST).build();
    }

    @DELETE
    @Path("/{idInvitacion}")
    public Response eliminarInvitacion(@PathParam("idInvitacion") String idInvitacion,
                                        @Valid @NotNull Invitacion invitacion){
        boolean seElimino = daoInvitacion.eliminarInvitacion(invitacion, idInvitacion);
        return seElimino ?
                Response.ok().build() :
                Response.status(Response.Status.BAD_REQUEST).build();
    }
}

File: src/main/java/rest/InvitacionResource.kt
package rest

import annotation.filter.ServerEventBroadcaster
import entity.Error
import entity.Invitacion
import org.hibernate.validator.constraints.NotEmpty
import usecase.InvitacionUseCase
import javax.validation.Valid
import javax.validation.constraints.NotNull
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/invitaciones")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
class InvitacionResource(private val invitacionUseCase: InvitacionUseCase) {

    @POST
    @ServerEventBroadcaster
    fun hacerInvitacion(@Valid @NotNull invitacion: Invitacion): Response {
        val resultado = invitacionUseCase.hacerInvitacion(invitacion)
        return when(resultado){
            is Error -> Response.status(Response.Status.BAD_REQUEST).entity(resultado).build()
            else -> Response.ok(resultado).build()
        }
    }

    @PUT
    @Path("/{idInvitacion}")
    @ServerEventBroadcaster
    fun responderInvitacion(@PathParam("idInvitacion") idInvitacion: String,
                            @Valid @NotNull invitacion: Invitacion,
                            @QueryParam("aceptar") @NotNull aceptar: Boolean): Response {
        val resultado: Any
        if (aceptar) {
            resultado = invitacionUseCase.aceptarInvitacion(invitacion, idInvitacion)
        } else {
            resultado = invitacionUseCase.rechazarInvitacion(invitacion, idInvitacion)
        }
        return when(resultado){
            is Error -> Response.status(Response.Status.BAD_REQUEST).entity(resultado).build()
            else -> Response.ok(resultado).build()
        }
    }

    @DELETE
    @Path("/{idInvitacion}")
    fun eliminarInvitacion(@PathParam("idInvitacion") idInvitacion: String): Response {
        val resultado = invitacionUseCase.eliminarInvitacion(idInvitacion)
        return when(resultado){
            is Error -> Response.status(Response.Status.BAD_REQUEST).entity(resultado).build()
            else -> Response.ok(resultado).build()
        }
    }
}